From c90c68cf6ff36543c99eca72618fe741642d8b6b Mon Sep 17 00:00:00 2001
From: NelloKudo <marshnelloosu@gmail.com>
Date: Mon, 5 Jan 2026 17:04:43 +0100
Subject: [PATCH 01/11] mmdevapi: Allow loading winealsa with
 PROTON_USE_WINEALSA=1.

---
 dlls/mmdevapi/main.c | 12 ++++++++++++
 1 file changed, 12 insertions(+)

diff --git a/dlls/mmdevapi/main.c b/dlls/mmdevapi/main.c
index 5426870e7a3..448aa5740c8 100644
--- a/dlls/mmdevapi/main.c
+++ b/dlls/mmdevapi/main.c
@@ -131,6 +131,7 @@ static BOOL WINAPI init_driver(INIT_ONCE *once, void *param, void **context)
     DriverFuncs driver;
     HKEY key;
     WCHAR reg_list[256], *p, *next, *driver_list = default_list;
+    WCHAR env_use_alsa[2];
 
     if(RegOpenKeyW(HKEY_CURRENT_USER, drv_keyW, &key) == ERROR_SUCCESS){
         DWORD size = sizeof(reg_list);
@@ -148,6 +149,17 @@ static BOOL WINAPI init_driver(INIT_ONCE *once, void *param, void **context)
         RegCloseKey(key);
     }
 
+    if (GetEnvironmentVariableW(L"PROTON_USE_WINEALSA", env_use_alsa, 2) > 0)
+    {
+        if (_wtoi(env_use_alsa) == 1)
+        {
+            default_list[0] = L'\0';
+            wcscat(default_list, L"alsa");
+            driver_list = default_list;
+            TRACE( "PROTON_USE_WINEALSA set, using alsa driver\n" );
+        }
+    }
+
     TRACE("Loading driver list %s\n", wine_dbgstr_w(driver_list));
     for(next = p = driver_list; next; p = next + 1){
         next = wcschr(p, ',');
-- 
2.54.0

