From bad28386ca0cc5200af3e0b35162d49f07d05c82 Mon Sep 17 00:00:00 2001
From: Ziia Shi <mkrsym1@gmail.com>
Date: Tue, 20 Jan 2026 16:33:26 +0100
Subject: [PATCH 02/11] HACK: kernel32: Spoof GetProcAddress of
 KiUserApcDispatcher and KiUserCallbackDispatcher.

---
 dlls/kernel32/module.c | 19 +++++++++++++++++++
 1 file changed, 19 insertions(+)

diff --git a/dlls/kernel32/module.c b/dlls/kernel32/module.c
index ed15e57c29c..54cf01e04ff 100644
--- a/dlls/kernel32/module.c
+++ b/dlls/kernel32/module.c
@@ -262,6 +262,17 @@ BOOL WINAPI GetBinaryTypeA( LPCSTR lpApplicationName, LPDWORD lpBinaryType )
     return GetBinaryTypeW(NtCurrentTeb()->StaticUnicodeString.Buffer, lpBinaryType);
 }
 
+#ifdef __x86_64__
+/* workaround for tpshell */
+static void __attribute__((naked)) int3_stub( void )
+{
+    asm("int3\t\n"
+        "int3\t\n"
+        "int3\t\n"
+        "int3\t\n");
+}
+#endif
+
 /***********************************************************************
  *           GetProcAddress   		(KERNEL32.@)
  *
@@ -285,6 +296,14 @@ FARPROC get_proc_address( HMODULE hModule, LPCSTR function )
     {
         ANSI_STRING     str;
 
+#ifdef __x86_64__
+        if (strcmp( function, "KiUserApcDispatcher" ) == 0 || strcmp( function, "KiUserCallbackDispatcher" ) == 0)
+        {
+            FIXME( "HACK: returning int3 stub instead of %s\n", function );
+            return (FARPROC)&int3_stub;
+        }
+#endif
+
         RtlInitAnsiString( &str, function );
         if (!set_ntstatus( LdrGetProcedureAddress( hModule, &str, 0, (void**)&fp ))) return NULL;
     }
-- 
2.54.0

