From 615f3bdcecef536375359f7dc17e0e7da7744430 Mon Sep 17 00:00:00 2001
From: Dylan Donnell <dylan.donnell@student.griffith.ie>
Date: Sun, 19 Oct 2025 13:37:58 +0200
Subject: [PATCH 11/11] ntdll: Stop unwinding on access violation.

---
 dlls/ntdll/signal_x86_64.c | 37 +++++++++++++++++++++++++++----------
 1 file changed, 27 insertions(+), 10 deletions(-)

diff --git a/dlls/ntdll/signal_x86_64.c b/dlls/ntdll/signal_x86_64.c
index 3237027b28c..ff9e5d9846a 100644
--- a/dlls/ntdll/signal_x86_64.c
+++ b/dlls/ntdll/signal_x86_64.c
@@ -881,6 +881,13 @@ __ASM_GLOBAL_FUNC( RtlRaiseException,
                    "movq %rax,%rcx\n\t"
                    "call " __ASM_NAME("RtlRaiseStatus") /* does not return */ );
 
+static LONG WINAPI walk_frame_chain_handler( EXCEPTION_POINTERS *eptr )
+{
+    if (eptr->ExceptionRecord->ExceptionCode == STATUS_ACCESS_VIOLATION)
+        return EXCEPTION_EXECUTE_HANDLER;
+
+    return EXCEPTION_CONTINUE_SEARCH;
+}
 
 /*************************************************************************
  *		RtlGetNativeSystemInformation (NTDLL.@)
@@ -923,18 +930,28 @@ ULONG WINAPI RtlWalkFrameChain( void **buffer, ULONG count, ULONG flags )
 
     RtlCaptureContext( &context );
 
-    for (i = 0; i < count; i++)
+    __TRY
     {
-        func = RtlLookupFunctionEntry( context.Rip, &base, &table );
-        if (!func) break;
-        if (RtlVirtualUnwind2( UNW_FLAG_NHANDLER, base, context.Rip, func, &context, NULL,
-                               &data, &frame, NULL, NULL, NULL, &handler, 0 ))
-            break;
-        if (!context.Rip) break;
-        if (!frame || !is_valid_frame( frame )) break;
-        if (context.Rsp == (ULONG_PTR)NtCurrentTeb()->Tib.StackBase) break;
-        if (i >= skip) buffer[num_entries++] = (void *)context.Rip;
+        for (i = 0; i < count; i++)
+        {
+            func = RtlLookupFunctionEntry( context.Rip, &base, &table );
+            if (!func) break;
+            if (RtlVirtualUnwind2( UNW_FLAG_NHANDLER, base, context.Rip, func, &context, NULL,
+                                   &data, &frame, NULL, NULL, NULL, &handler, 0 ))
+                break;
+            if (!context.Rip) break;
+            if (!frame || !is_valid_frame( frame )) break;
+            if (context.Rsp == (ULONG_PTR)NtCurrentTeb()->Tib.StackBase) break;
+            if (i >= skip) buffer[num_entries++] = (void *)context.Rip;
+        }
+    }
+    __EXCEPT(walk_frame_chain_handler)
+    {
+        TRACE( "access violation when unwinding with context.Rip %p\n", (void *)context.Rip );
+        return 0;
     }
+    __ENDTRY
+
     return num_entries;
 }
 
-- 
2.54.0

